### ISSUE TITLE
CleanupReporter does not error if tests fail

### ISSUE THREAD

**Issue by jeroen** [r-lib/ps#163]
The av package uses this:

```r
reporter <- ps::CleanupReporter(testthat::CheckReporter)$new(proc_cleanup = FALSE, proc_fail = FALSE, conn_fail = FALSE)
test_check("av", reporter = reporter)
```

However test failures have gone unnoticed because this does not error if the tests fail. 

For example see: https://github.com/ropensci/av/actions/runs/10263707793 (you can see there is a test error for macos-latest in the GHA annotations or artifacts but the check still says ✅ )

**Comment by gaborcsardi** [comment]
Simpler reprex, with ps itself, this fails as it should:
```r
testthat::test_local(reporter = "summary", stop_on_failure = TRUE, filter = "wait")
```

But this does not:
```r
reporter <- ps::CleanupReporter(testthat::SummaryReporter)$new()
testthat::test_local(reporter = reporter, stop_on_failure = TRUE, filter = "wait")
```

**Comment by gaborcsardi** [comment]
We could easily fix this in testthat, but then nothing happens until a testthat release. I'll see if I can work around it somehow...

**Comment by gaborcsardi** [comment]
For the record, the issue is that testthat only looks at the last expectation when it decides whether there was an error in a `test_that()` block. But with the cleanup reporter adding extra expectations, the error expectation is not necessarily the last one any more. This patch for testthat fixes the issue:
```patch
diff --git a/R/reporter-list.R b/R/reporter-list.R
index 39b0c3d7..1c90515d 100644
--- a/R/reporter-list.R
+++ b/R/reporter-list.R
@@ -168,10 +168,10 @@ summarize_one_test_results <- function(test) {
   if (nb_tests > 0) {
     # error reports should be handled differently.
     # They may not correspond to an expect_that() test so remove them
-    last_test <- test_results[[nb_tests]]
-    error <- expectation_error(last_test)
+    error_results <- map_lgl(test_results, expectation_error)
+    error <- length(error_results) > 0
     if (error) {
-      test_results <- test_results[-nb_tests]
+      test_results <- test_results[-error_results]
       nb_tests <- length(test_results)
     }
```

**Comment by gaborcsardi** [comment]
One idea is to put in the cleanup expectations immediately when an error expectation comes in, because that's surely the end of the `test_that()` block. 

The problem with that is that it happens before `on.exit()` runs, so we might check for leftover connections before the potential cleanup runs.

**Comment by gaborcsardi** [comment]
A not fully correct, but still better workaround is to not put in the cleanup expectations at all, unless they fail. This would keep the error expectation the last.

Maybe we can do this with the current testthat version, and also do the testthat fix, so we don't have to to do it with future versions.

**Comment by gaborcsardi** [comment]
Actually, testthat (or at least the testthat testsuite) very heavily relies on the current behavior, so I'll work around this completely in ps.

We are not going to add expectations or even check for leftover connections, etc. if the `test_that()` block had an error.

**Comment by gaborcsardi** [comment]
Closed by #169.
